### ISSUE TITLE
Issue with running 'sudo geneos init' 

### ISSUE THREAD

**Issue by gvastel** [ITRS-Group/cordial#8]
Version v1.3.1, cordial / tools / geneos /

When running `sudo geneos init itrs /opt/itrs`, the directory structure under `/opt/itrs` is created, but not the configuration file `/home/itrs/.config/geneos/geneos.json`.
In this case, the configuration file is also not found in the root user's home directory.

When running `geneos init /opt/itrs` as user `itrs`, both the directory structure under `/opt/itrs` & the configuration file `/home/itrs/.config/geneos/geneos.json` are created.

Looks like there is an issue with the command run as sudo.

**Comment by itrs-dsobiepan** [comment]
I'm going to tack on that I had the same type of issue occur with the .config file creation when setting up in an installation directory I already had:

```
[dsobiepan@nysupsvr21 bin]$ ./geneos set dsobiepan geneos=/export/home/dsobiepan/opt
Error:
Geneos installation directory not set.

Use one of the following to fix this:

For an existing installation:
        $ geneos set user geneos=/path/to/geneos

To initialise a new installation:
        $ geneos init /path/to/geneos

For temporary usage:
        $ export ITRS_HOME=/path/to/geneos

[dsobiepan@nysupsvr21 bin]$ ./geneos set user dsobiepan=/export/home/dsobiepan/opt
Error: open /export/home/dsobiepan/.config/geneos/geneos.json: no such file or directory
```

I think both methods doesn't generate the /.config directory and geneos.json file. When I try to manually create them, I end up getting a fatal issue when rerunning the set command.

**Comment by pgalbavy-itrs** [comment]
@itrs-dsobiepan in your case the issue is one of poor / unclear documentation. You sghould use the literal work `user` and not your account name, e.g.

`./geneos set user geneos=/export/home/dsobiepan/opt`

The second example you tried is also not quite a known syntax.

The sub-sub commands are `geneos set global` and `geneos set user` and then the key=value part is `geneos` and the path

Will take that into account when updating docs.

**Comment by itrs-dsobiepan** [comment]
Thanks Peter appreciate you clarifying.

Unfortunately still hitting the same wall with the corrected syntax:

```
[dsobiepan@nysupsvr21 bin]$ ./geneos set user geneos=/export/home/dsobiepan/opt
Error: open /export/home/dsobiepan/.config/geneos/geneos.json: no such file or directory
[dsobiepan@nysupsvr21 bin]$ ./geneos set user geneos=/export/home/dsobiepan/opt/
Error: open /export/home/dsobiepan/.config/geneos/geneos.json: no such file or directory
```
Am I missing some other step? Maybe it's something to do with the environment I'm on? For context, this is on the NY VMs we have in the office, so plenty of other users alongside me

**Comment by pgalbavy-itrs** [comment]
> When running `sudo geneos init itrs /opt/itrs`, the directory structure under `/opt/itrs` is created, but not the configuration file `/home/itrs/.config/geneos/geneos.json`. In this case, the configuration file is also not found in the root user's home directory.

I think this is because the original code writes a global config file:

```go
		if utils.IsSuperuser() {
			if err = host.WriteConfigFile(GlobalConfigPath, "root", 0664, config.GetConfig()); err != nil {
				log.Fatal().Err(err).Msg("cannot write global config")
			}
		} else {
			userConfFile := UserConfigFilePaths()[0]

			if err = host.WriteConfigFile(userConfFile, opts.localusername, 0664, config.GetConfig()); err != nil {
				return err
			}
		}
```

This code path is from way-back-when and is probably not relevent anymore. I think we can just reduce this down to writing in the target user's config directory.

**Comment by pgalbavy-itrs** [comment]
> Am I missing some other step? Maybe it's something to do with the environment I'm on? For context, this is on the NY VMs we have in the office, so plenty of other users alongside me

Can you run this with a '-d' flag and post the last chunk of lines that may be relevant. It's not from trying to write the file out as that actually doesn't return errors (something else to fix).

**Comment by itrs-dsobiepan** [comment]
```
[dsobiepan@nysupsvr21 bin]$ ./geneos set user geneos=/export/home/dsobiepan/opt/ -d
2022-11-02T12:16:57-04:00 DEBUG: tools/geneos/cmd/parseargs.go:47 cmd.parseArgs() > rawargs: [geneos=/export/home/dsobiepan/opt/]
2022-11-02T12:16:57-04:00 DEBUG: tools/geneos/cmd/parseargs.go:63 cmd.parseArgs() > rawargs [], params [geneos=/export/home/dsobiepan/opt/], ct
2022-11-02T12:16:57-04:00 DEBUG: tools/geneos/cmd/parseargs.go:214 cmd.cmdArgs() > map[args:[] ct: params:["geneos=/export/home/dsobiepan/opt/"] wildcard:false] <nil>
Error: open /export/home/dsobiepan/.config/geneos/geneos.json: no such file or directory
```

It turns out the .config directory exists / was created, but the geneos directory and its .json file were not:

```
[dsobiepan@nysupsvr21 bin]$ ls -al /export/home/dsobiepan/.config/
total 12
drwxrwxr-x.  2 dsobiepan dsobiepan 4096 Nov  2 11:42 .
drwxrwxrwx. 22 dsobiepan dsobiepan 4096 Nov  2 11:42 ..

```

**Comment by itrs-dsobiepan** [comment]
Thanks Peter!

**Comment by pgalbavy-itrs** [comment]
> Thanks Peter!

However, I hate git/github. If I cannot figure out how to use cherry-pick to pull the change into main and then roll a patch release you may have to wait for v1.4.0 coming soon.

**Comment by pgalbavy-itrs** [comment]
Most of these issues (not all) are fixed in release v1.3.2 which is just being uploaded.

**Comment by gvastel** [comment]
v1.3.2 still does not create `/home/itrs/.config/geneos/geneos.json` when running `sudo geneos init itrs /opt/itrs`.
This was used on and AWS VM running Centos7.

**Comment by pgalbavy-itrs** [comment]
Having review this, I think it needs (yet) more options. There are at least conflicting two scenarios:

1. Running with `sudo` to provide the power to create base directory but then saving configuration details in target, non-root, user's configuration directory and files
2. Running with `sudo` in expectation of doing so in the future, e.g. setting the configuration in the root user's home area or the global config file, so that doing `sudo geneos start` picks up the correct base directory which includes the `defaultuser` setting etc.

In this case I am uncertain how to have the user express this on the command like, unless we add more flags like `--as-user` and `--for-user` picking one as the default?

I want to stick to the Principle of Least Astonishment, but different users will have different expectations.

Ideas?

**Comment by gvastel** [comment]
When running `sudo geneos init`, you need to explicitly define the user (1st argument) as in the geneos user who will own the setup & run geneos commands.
In this case, for scenario (2) above, that user should be `root` and config files placed in  the `root` user's home directory (`/home/itrs/.config/geneos/geneos.json`), and the geneos base directory should be set to what is defined (or the default assuming the user is `root`).
And for scenarios (1), permissions from sudo can be used to create files & directories to be owner by the defined user (e.g. `itrs`).
In other words the `as-user` should always be the user running the geneos command.  With `sudo` that'd be `root`, and without sudo that would be the current user, e.g. `itrs`.  And the `as-user` should be the user defined as 1st argument in the `geneos init` command; that parameter could be made mandatory to avoid any doubt.

**Comment by pgalbavy-itrs** [comment]
The user handling in `geneos` is somewhat haphazard and I intend to remove all of it and start again - v1.5.0 will just run vanilla instances and warn if the user in an instance file is not the same as the one running the program.

This will also address the discrepancy for remote hosts, where no priv elevation is supported.

The one area that support for `sudo` is retained is for `init` commands which check if the program us run under `sudo` (environment variables show this) and then creating directories as the underlying caller unless told otherwise. The code that tries to create files or execute binaries as a 3rd party is a mess and has to go. For now.

**Comment by pgalbavy-itrs** [comment]
Special handling of users is mostly gone now. Once the new version settles we can look at doing it again, properly.

Closing this for now.
